tools/test-runner/src/main.rs

```diff
@@ -0,0 +1,196 @@
+            }
+            if test.contacts.len() > 0 {
+                let contacts = format!("Please report this to {:?}\n", test.contacts);
+                message.push_str(&contacts);
+            }
+            errors.push(message);
+        }
+    }
+    for error in errors {
+        eprintln!("{}", error);
+    }
+    let failed = total - passed;
+    println!("Out of {} tests, {} passed, {} failed", total, passed, failed);
+
+    // generate test report
+    match options.report {
+        None => (),
+        Some(path) => report(tests, path)?,
+    }
+
+    exit(failed as i32);
+}
+
+// Refer to https://llg.cubic.org/docs/junit/
+fn report(tests: Vec<TestResult>, path: PathBuf) -> Result<()> {
+    let file = File::create(&path).with_context(|| {
+        format!(
+            "Couldn't generate report at specified path {}",
+            path.display()
+        )
+    })?;
+    let mut xml = BufWriter::new(&file);
+
+    let failures = tests.iter().filter(|test| !test.passed).count();
+    writeln!(xml, r#"<?xml version="1.0" encoding="UTF-8"?>"#)?;
+    writeln!(xml, r#"<testsuites tests="{}" failures="{}">"#, tests.len(), failures)?;
+
+    // we group unit tests from the same buck target as a JUnit "testsuite"
+    let suites = tests
+        .into_iter()
+        .map(|test| {
+            let name: Vec<&str> = test.name.split("#").collect();
+            let target = name[0].to_owned();
+            let unit = if name.len() > 1 {
+                name[1].to_owned()
+            } else {
+                target.clone()
+            };
+            return (target, unit, test);
+        })
+        .into_group_map_by(|(target, _, _)| target.to_owned());
+    for (target, cases) in suites {
+        let failures = cases.iter().filter(|(_, _, test)| !test.passed).count();
+        writeln!(xml, r#"  <testsuite name="{}" tests="{}" failures="{}">"#,
+                      target, cases.len(), failures)?;
+
+        for (target, unit, test) in cases {
+            write!(xml, r#"    <testcase classname="{}" name="{}" time="{}""#,
+                        target, unit, test.duration.as_millis() as f32 / 1e3)?;
+            if test.passed {
+                writeln!(xml, " />")?;
+            } else {
+                writeln!(xml, r#">"#)?;
+                writeln!(xml, r#"      <failure>Test failed after {} unsuccessful attempts</failure>"#, test.attempts)?;
+                writeln!(xml, r#"      <system-out>{}</system-out>"#,
+                              xml_escape_text(test.stdout))?;
+                writeln!(xml, r#"      <system-err>{}</system-err>"#,
+                              xml_escape_text(test.stderr))?;
+                writeln!(xml, r#"    </testcase>"#)?;
+            }
+        }
+
+        writeln!(xml, r#"  </testsuite>"#)?;
+    }
+
+    writeln!(xml, r#"</testsuites>"#)?;
+
+    eprintln!("Test report written to {}", path.display());
+    return Ok(());
+}
+
+fn xml_escape_text(unescaped: String) -> String {
```

Comment:
This is fine for now, but may end up causing unexpected issues in the future. Don't fix it in this PR, but `serde_xml_rs` may work (https://docs.rs/serde-xml-rs/0.5.0/serde_xml_rs/) - if not (there are some things that it can't serialize) then I have used [quick_xml](https://docs.rs/quick-xml/0.22.0/quick_xml/) in the past

Reply:
Yeah since I had a very specific format to generate, I thought doing it by hand would be easier.
Later on I realized we would need to escape stdout/err, so using a dedicated library would probably be less error prone.